File: extensions-api/src/main/scala/pl/touk/nussknacker/engine/api/process/ProcessConfigCreator.scala
package pl.touk.nussknacker.engine.api.process

import pl.touk.nussknacker.engine.ModelConfig
import pl.touk.nussknacker.engine.api.{CustomStreamTransformer, ProcessListener, Service}
import pl.touk.nussknacker.engine.api.modelinfo.ModelInfo

/**
  * There Nussknacker fetches information about user defined model.
  * Any invocation of user defined logic or resource goes through this class.
  */
trait ProcessConfigCreator extends Serializable {

  def customStreamTransformers(
      modelConfig: ModelConfig
  ): Map[String, WithCategories[CustomStreamTransformer]]

  def services(modelConfig: ModelConfig): Map[String, WithCategories[Service]]

  def sourceFactories(modelConfig: ModelConfig): Map[String, WithCategories[SourceFactory]]

  def sinkFactories(modelConfig: ModelConfig): Map[String, WithCategories[SinkFactory]]

  def listeners(modelConfig: ModelConfig): Seq[ProcessListener]

  def expressionConfig(modelConfig: ModelConfig): ExpressionConfig

  def modelInfo(): ModelInfo

  def asyncExecutionContextPreparer(
      modelConfig: ModelConfig
  ): Option[AsyncExecutionContextPreparer] = None

  def classExtractionSettings(modelConfig: ModelConfig): ClassExtractionSettings =
    ClassExtractionSettings.Default

}

File: extensions-api/src/main/java/pl/touk/nussknacker/engine/javaapi/process/ProcessConfigCreator.java
package pl.touk.nussknacker.engine.javaapi.process;

import pl.touk.nussknacker.engine.ModelConfig;
import pl.touk.nussknacker.engine.api.CustomStreamTransformer;
import pl.touk.nussknacker.engine.api.ProcessListener;
import pl.touk.nussknacker.engine.api.Service;
import pl.touk.nussknacker.engine.api.process.AsyncExecutionContextPreparer;
import pl.touk.nussknacker.engine.api.process.SinkFactory;
import pl.touk.nussknacker.engine.api.process.SourceFactory;
import pl.touk.nussknacker.engine.api.process.WithCategories;

import java.io.Serializable;
import java.util.Collection;
import java.util.Map;
import java.util.Optional;

public interface ProcessConfigCreator extends Serializable {

    Map<String, WithCategories<Service>> services(ModelConfig modelConfig);
    Map<String, WithCategories<SourceFactory>> sourceFactories(ModelConfig modelConfig);
    Map<String, WithCategories<SinkFactory>> sinkFactories(ModelConfig modelConfig);
    Map<String, WithCategories<CustomStreamTransformer>> customStreamTransformers(ModelConfig modelConfig);
    Collection<ProcessListener> listeners(ModelConfig modelConfig);
    ExpressionConfig expressionConfig(ModelConfig modelConfig);
    Map<String, String> modelInfo();
    default Optional<AsyncExecutionContextPreparer> asyncExecutionContextPreparer(ModelConfig modelConfig) {
      return Optional.empty();
    }
    default ClassExtractionSettings classExtractionSettings(ModelConfig modelConfig) {
        return ClassExtractionSettings.DEFAULT;
    }

}
